ddht/v5_1/alexandria/client.py

```diff
@@ -261,6 +315,55 @@ async def ping(self, node_id: NodeID, endpoint: Endpoint) -> PongMessage:
         response = await self._request(node_id, endpoint, request, PongMessage)
         return response
 
+    async def find_nodes(
+        self,
+        node_id: NodeID,
+        endpoint: Endpoint,
+        distances: Collection[int],
+        *,
+        request_id: Optional[bytes] = None,
+    ) -> Tuple[InboundMessage[FoundNodesMessage], ...]:
+        request = FindNodesMessage(FindNodesPayload(tuple(distances)))
+
+        subscription: trio.abc.ReceiveChannel[InboundMessage[FoundNodesMessage]]
+        # unclear why `subscribe_request` isn't properly carrying the type information
+        async with self.subscribe_request(  # type: ignore
+            node_id, endpoint, request, FoundNodesMessage, request_id=request_id,
+        ) as subscription:
+            head_response = await subscription.receive()
+            total = head_response.message.payload.total
+            responses: Tuple[InboundMessage[FoundNodesMessage], ...]
+            if total == 1:
+                responses = (head_response,)
+            elif total > 1:
+                tail_responses: List[InboundMessage[FoundNodesMessage]] = []
+                for _ in range(total - 1):
+                    tail_responses.append(await subscription.receive())
```

Comment:
I know this is just copying the existing implementation of `find_nodes`, so maybe the place to fix this is in a different PR, but this seems like this is a little too pessimistic. If the remote node sends us three `FoundNodes` messages but the last one is dropped we'll time out and return nothing to the caller, even though we received 2 messages with useful enrs.

Reply:
Opening an issue to track this as well.  Probably something we want to measure in practice and see how often we encounter this behavior.